subprojects/worker-processes/src/main/java/org/gradle/process/internal/worker/messaging/WorkerConfigSerializer.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.process.internal.worker.messaging;
+
+import org.gradle.api.Action;
+import org.gradle.api.logging.LogLevel;
+import org.gradle.internal.io.ClassLoaderObjectInputStream;
+import org.gradle.internal.remote.internal.inet.MultiChoiceAddress;
+import org.gradle.internal.remote.internal.inet.MultiChoiceAddressSerializer;
+import org.gradle.internal.serialize.Decoder;
+import org.gradle.internal.serialize.Encoder;
+import org.gradle.internal.serialize.InputStreamBackedDecoder;
+import org.gradle.internal.serialize.OutputStreamBackedEncoder;
+import org.gradle.internal.serialize.Serializer;
+import org.gradle.process.internal.worker.WorkerProcessContext;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+
+/**
+ * Serializes and de-serializes {@link WorkerConfig}s.
+ */
+public class WorkerConfigSerializer implements Serializer<WorkerConfig> {
```

Comment:
Unit tests for this class seem like they would be a valuable benefit of extracting this login into a class.

Reply:
Wrote a test, though this project had no unit tests previously. I'm having a hard time getting `AbstractIntegrationSpec` properly accessible in this project (seems directly adding it as a dep is not the right way to do this.)

Reply:
With the newest commit, I get 
```

java.lang.ExceptionInInitializerError
	at org.gradle.integtests.fixtures.AbstractIntegrationSpec.createExecuter(AbstractIntegrationSpec.groovy:129)
	at org.gradle.integtests.fixtures.AbstractIntegrationSpec.getExecuter(AbstractIntegrationSpec.groovy:76)
	at org.gradle.integtests.fixtures.AbstractIntegrationSpec.setup(AbstractIntegrationSpec.groovy:104)
	Suppressed: java.lang.NoClassDefFoundError: Could not initialize class org.gradle.integtests.fixtures.executer.GradleContextualExecuter
		at org.gradle.integtests.fixtures.AbstractIntegrationSpec.createExecuter(AbstractIntegrationSpec.groovy:129)
		at org.gradle.integtests.fixtures.AbstractIntegrationSpec.getExecuter(AbstractIntegrationSpec.groovy:76)
		at org.gradle.integtests.fixtures.AbstractIntegrationSpec.cleanup(AbstractIntegrationSpec.groovy:114)
Caused by: org.gradle.internal.service.ServiceCreationException: Could not create service of type ClassLoaderRegistry using GlobalScopeServices.createClassLoaderRegistry().
	at org.gradle.internal.service.DefaultServiceRegistry$FactoryMethodService.invokeMethod(DefaultServiceRegistry.java:898)
	at org.gradle.internal.service.DefaultServiceRegistry$FactoryService.createServiceInstance(DefaultServiceRegistry.java:823)
	at org.gradle.internal.service.DefaultServiceRegistry$ManagedObjectServiceProvider.getInstance(DefaultServiceRegistry.java:608)
	at org.gradle.internal.service.DefaultServiceRegistry$SingletonService.get(DefaultServiceRegistry.java:671)
	at org.gradle.internal.service.DefaultServiceRegistry.applyConfigureMethod(DefaultServiceRegistry.java:199)
	at org.gradle.internal.service.DefaultServiceRegistry.findProviderMethods(DefaultServiceRegistry.java:181)
	at org.gradle.internal.service.DefaultServiceRegistry.addProvider(DefaultServiceRegistry.java:265)
	at org.gradle.internal.service.ServiceRegistryBuilder.build(ServiceRegistryBuilder.java:52)
	at org.gradle.integtests.fixtures.executer.AbstractGradleExecuter.<clinit>(AbstractGradleExecuter.java:111)
	... 3 more
Caused by: java.lang.NullPointerException
	at org.gradle.util.internal.GUtil.loadProperties(GUtil.java:242)
	at org.gradle.api.internal.classpath.DefaultPluginModuleRegistry.loadPluginProperties(DefaultPluginModuleRegistry.java:60)
	at org.gradle.api.internal.classpath.DefaultPluginModuleRegistry.getApiModules(DefaultPluginModuleRegistry.java:36)
	at org.gradle.api.internal.DynamicModulesClassPathProvider.gradleExtensionsWithout(DynamicModulesClassPathProvider.java:72)
	at org.gradle.api.internal.DynamicModulesClassPathProvider.findClassPath(DynamicModulesClassPathProvider.java:51)
	at org.gradle.api.internal.DefaultClassPathRegistry.getClassPath(DefaultClassPathRegistry.java:35)
	at org.gradle.initialization.DefaultClassLoaderRegistry.<init>(DefaultClassLoaderRegistry.java:35)
	at org.gradle.internal.service.scopes.GlobalScopeServices.createClassLoaderRegistry(GlobalScopeServices.java:295)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:125)
	at org.gradle.internal.service.ReflectionBasedServiceMethod.invoke(ReflectionBasedServiceMethod.java:34)
	at org.gradle.internal.service.DefaultServiceRegistry$FactoryMethodService.invokeMethod(DefaultServiceRegistry.java:896)
	... 11 more
```

Reply:
There's no need to extend `AbstractIntegrationSpec` for a unit test, only if you want to launch Gradle within your test.  Just extend `Specification` and the test will run.  Looks like there are some issues with the `workerAction` not being `Serializable` though.